collector/exporter/otelarrowexporter/internal/arrow/stream.go

```diff
@@ -197,83 +207,24 @@ func (s *Stream) run(bgctx context.Context, streamClient StreamClientFunc, grpcO
-			case codes.Unimplemented:
-				// This (client == nil) signals the controller
-				// to downgrade when all streams have returned
-				// in that status.
-				//
-				// TODO: Note there are partial failure modes
-				// that will continue to function in a
-				// degraded mode, such as when half of the
-				// streams are successful and half of streams
-				// take this return path.  Design a graceful
-				// recovery mechanism?
-				s.client = nil
-				s.telemetry.Logger.Info("arrow is not supported",
-					zap.String("message", status.Message()),
-				)
-
-			case codes.Unavailable, codes.Internal:
-				// gRPC returns this when max connection age is reached.
-				// The message string will contain NO_ERROR if it's a
-				// graceful shutdown.
-				//
-				// Having seen:
-				//
-				//     arrow stream unknown {"kind": "exporter",
-				//     "data_type": "traces", "name": "otlp/traces",
-				//     "code": 13, "message": "stream terminated by
-				//     RST_STREAM with error code: NO_ERROR"}
-				//
-				// from the default case below print `"code": 13`, this
-				// branch is now used for both Unavailable (witnessed
-				// in local testing) and Internal (witnessed in
-				// production); in both cases "NO_ERROR" is the key
-				// signifier.
-				if strings.Contains(status.Message(), "NO_ERROR") {
-					s.telemetry.Logger.Error("arrow stream reset (consider lowering max_stream_lifetime)",
-						zap.String("message", status.Message()),
-					)
-				} else {
-					s.telemetry.Logger.Error("arrow stream unavailable",
-						zap.String("message", status.Message()),
-					)
-				}
-
-			case codes.Canceled:
-				// Note that when the writer encounters a local error (such
-				// as a panic in the encoder) it will cancel the context and
-				// writeErr will be set to an actual error, while the error
-				// returned from read() will be the cancellation by the
-				// writer. So if the reader's error is canceled and the
-				// writer's error is non-nil, use it instead.
-				if writeErr != nil {
-					s.telemetry.Logger.Error("arrow stream internal error",
-						zap.Error(writeErr),
-					)
-					// reset the writeErr so it doesn't print below.
-					writeErr = nil
-				} else {
-					s.telemetry.Logger.Error("arrow stream canceled",
-						zap.String("message", status.Message()),
-					)
-				}
-			default:
-				s.telemetry.Logger.Error("arrow stream unknown",
-					zap.Uint32("code", uint32(status.Code())),
-					zap.String("message", status.Message()),
-				)
-			}
+		if status, ok := status.FromError(err); ok && status.Code() == codes.Unimplemented {
+			// This (client == nil) signals the controller to
+			// downgrade when all streams have returned in that
+			// status.
+			//
+			// This is a special case because we reset s.client,
+			// which sets up a downgrade after the streams return.
+			s.client = nil
+			s.telemetry.Logger.Info("arrow is not supported",
+				zap.String("message", status.Message()),
+			)
```

Comment:
Note this is a special case carried over from the old code. The rest of this rewrite in this function body is a major simplification.